MetaMask/metamask-extension
# Converting some ButtonIcon files to TS and creating d.ts file

## Explanation

Currently the `ButtonIcon` component is written in JavaScript. This PR migrates some of the files to TS and creates a `d.ts` types file so the `ButtonIcon` can be used in other ts files. It will unblock the following PRs:

- https://github.com/MetaMask/metamask-extension/pull/17097
- https://github.com/MetaMask/metamask-extension/pull/18311
- https://github.com/MetaMask/metamask-extension/pull/18156

* Part of #18327

## Screenshots/Screencaps

### Before

Linting issues that will fail in `yarn lint` test

![Screenshot 2023-04-05 at 11 01 36 AM](https://user-images.githubusercontent.com/8112138/230165700-c89050fc-6149-4df8-bf7d-d2fa39d2304a.png)

No changes visually to the `ButtonIcon` component in storybook


https://user-images.githubusercontent.com/8112138/230137790-79ca8115-26ca-4647-a5ed-ab397d002501.mov

### After

![Screenshot 2023-04-05 at 11 02 21 AM](https://user-images.githubusercontent.com/8112138/230165764-e9450c41-4adb-46e9-b4e6-34fb9ee5f51a.png)


No changes visually to the `ButtonIcon` component in storybook


https://user-images.githubusercontent.com/8112138/230154690-751e061c-ebb2-418c-b117-1e71efc0b76c.mov

## Manual Testing Steps

- Go to the storybook build on this PR
- Search `ButtonIcon` 
- Check controls, stories and docs
To check types

- Go to `ui/components/component-library/header-base/header-base.stories.tsx`
- Find `<ButtonIcon>` component
- See TS issue

## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
